<a href="https://colab.research.google.com/github/Yuxi2910/SQL_Project1_Instacart/blob/main/sqlWitchcraft.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
# install ipython-sql
!pip install ipython-sql

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.6/1.6 MB 17.9 MB/s eta 0:00:00


In [2]:
%load_ext sql

In [4]:
%pip install mysql pymysql

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 91.4/91.4 kB 3.1 MB/s eta 0:00:00
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 45.3/45.3 kB 2.9 MB/s eta 0:00:00
  Created wheel for mysqlclient: filename=mysqlclient-2.2.7-cp312-cp312-linux_x86_64.whl size=129069 sha256=646883c432a67914a95ceb751801498899261c0e93dafadcc31028948b69b097
  Stored in directory: /root/.cache/pip/wheels/27/95/18/7f176fffd46629e710c04c810b9c4d7d4358fe7c96a7d2306d
Successfully built mysqlclient


In [6]:
import sqlalchemy
engine = sqlalchemy.create_engine('mysql+pymysql://student:guest@columbia-mds.csbmzoea3lu9.us-east-1.rds.amazonaws.com:3306/world')
engine.connect()

In [10]:
!pip install jupysql

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 95.1/95.1 kB 2.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 192.8/192.8 kB 9.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 145.2/145.2 kB 11.3 MB/s eta 0:00:00


In [1]:
%load_ext sql

## Q1. Getting to know the data

In [3]:
# a) get all tables in the database
%sql mysql+pymysql://student:guest@columbia-mds.csbmzoea3lu9.us-east-1.rds.amazonaws.com:3306/witchcraft

Connecting and switching to connection 'mysql+pymysql://student:***@columbia-mds.csbmzoea3lu9.us-east-1.rds.amazonaws.com:3306/witchcraft'

In [4]:
%sql student@witchcraft

In [5]:
%sql SHOW TABLES

Running query in 'mysql+pymysql://student:***@columbia-mds.csbmzoea3lu9.us-east-1.rds.amazonaws.com:3306/witchcraft'

37 rows affected.

Tables_in_witchcraft
accused
accused_family
appeal
calendarcustom
case_person
commission
complaint
confession
counterstrategy
demonicpact


In [23]:
# b) Show the first three entries in the table accused
%%sql
SELECT * FROM accused LIMIT 3;

Running query in 'mysql+pymysql://student:***@columbia-mds.csbmzoea3lu9.us-east-1.rds.amazonaws.com:3306/witchcraft'

3 rows affected.

accusedref,accusedsystemid,accusedid,firstname,lastname,m_firstname,m_surname,alias,patronymic,destitle,sex,age,age_estcareer,age_estchild,res_settlement,res_parish,res_presbytery,res_county,res_burgh,res_ngr_letters,res_ngr_easting,res_ngr_northing,ethnic_origin,maritalstatus,socioecstatus,occupation,notes,createdby,createdate,lastupdatedby,lastupdatedon
A/EGD/10,EGD,10,Mareon,Quheitt,Marion,White,None,None,None,Female,None,0,0,Sammuelston,P/JO/3539,Haddington,Haddington,None,None,None,None,None,None,None,None,None,SMD,2001-05-15T11:06:51,jhm,2002-08-09T11:40:51
A/EGD/100,EGD,100,Thom,Cockburn,Thomas,Cockburn,None,None,None,Male,None,0,0,None,None,None,Haddington,None,None,None,None,None,None,None,None,None,SMD,2001-05-15T11:06:51,jhm,2002-10-02T10:32:30
A/EGD/1000,EGD,1000,Christian,Aitkenhead,Christine,Aikenhead,None,None,None,Female,None,0,0,Rottinraw,None,None,Dumfries,None,None,None,None,None,Married,None,None,None,SMD,2001-05-15T11:06:51,jhm,2002-10-01T10:48:12


In [24]:
# SELECT
%sql
SELECT firstname from accused;

Running query in 'mysql+pymysql://student:***@columbia-mds.csbmzoea3lu9.us-east-1.rds.amazonaws.com:3306/witchcraft'

3219 rows affected.

firstname
Mareon
Thom
Christian
Janet
Agnes
Margaret
Margaret
Janet
Agnes
Marion


In [18]:
# c) How many people are included in the accused table?
%%sql
SELECT COUNT(*) FROM accused;

Running query in 'mysql+pymysql://student:***@columbia-mds.csbmzoea3lu9.us-east-1.rds.amazonaws.com:3306/witchcraft'

1 rows affected.

COUNT(*)
3219


In [27]:
# d) Looks like the age is missing for some observations. how to count the number of non-missing values for age in the data.
%%sql
SELECT COUNT(age) FROM accused;

Running query in 'mysql+pymysql://student:***@columbia-mds.csbmzoea3lu9.us-east-1.rds.amazonaws.com:3306/witchcraft'

1 rows affected.

COUNT(age)
166


In [70]:
%config SqlMagic.displaylimit = None

displaylimit: Value None will be treated as 0 (no limit)

In [87]:
# e) Show a list of unique occupations.
%%sql
SELECT DISTINCT occupation
FROM accused;

Running query in 'mysql+pymysql://student:***@columbia-mds.csbmzoea3lu9.us-east-1.rds.amazonaws.com:3306/witchcraft'

32 rows affected.

occupation
None
Servant
Vagabond
Weaver
Midwife
Tailor
Messenger
Brewster
Smith
Minister


In [90]:
# f) What proportion of accused are female?
%%sql
SELECT ROUND((SUM(sex = 'Female') / COUNT(*)) *100, 2) AS 'Female_proportion'
FROM accused;

Running query in 'mysql+pymysql://student:***@columbia-mds.csbmzoea3lu9.us-east-1.rds.amazonaws.com:3306/witchcraft'

1 rows affected.

Female_proportion
83.94


83.94% suggests that a lot more females were being accused of witchcraft than males.

In [59]:
# g) Using CASE WHEN, categorize accused individuals by age groups: 'Young' (age < 30),
# 'Middle-aged' (age between 30 and 60, inclusive), and 'Old' (age > 60). Count how many
# individuals fall into each category, separately for males and females. Exclude missing
# age values from your analysis. Your result should show age category, sex, and count.

%%sql
SELECT
  CASE
    WHEN age < 30 THEN 'Young'
    WHEN age BETWEEN 30 AND 60 THEN 'Middle-aged'
    ELSE 'Old'
  END AS age_category,
  COUNT(*) AS accused_count,
  sex
FROM accused
GROUP BY age_category, sex;



Running query in 'mysql+pymysql://student:***@columbia-mds.csbmzoea3lu9.us-east-1.rds.amazonaws.com:3306/witchcraft'

7 rows affected.

age_category,accused_count,sex
Old,2571,Female
Old,444,Male
Middle-aged,111,Female
Young,20,Female
Middle-aged,19,Male
Old,49,None
Young,5,Male


Regardless of sex, old-age people are more likely to get accused of witchcraft, and middle-aged are more likely to get accused than young-aged. Overall, more females are accused than males for all age categories.

## Q2. Seeing the Devil
Let's look at some appearances of the devil in the devilappearance table.

In [61]:
# a) List the unique devil_types in the data.
%%sql
DESCRIBE devilappearance;

Running query in 'mysql+pymysql://student:***@columbia-mds.csbmzoea3lu9.us-east-1.rds.amazonaws.com:3306/witchcraft'

10 rows affected.

Field,Type,Null,Key,Default,Extra
devilref,text,YES,,None,
devilsystemid,text,YES,,None,
devilid,bigint,YES,,None,
caseref,text,YES,,None,
devil_type,text,YES,,None,
devil_text,text,YES,,None,
createdby,text,YES,,None,
createdate,text,YES,,None,
lastupdatedby,text,YES,,None,
lastupdatedon,text,YES,,None,


In [62]:
%%sql
SELECT *
FROM devilappearance
LIMIT 3;

Running query in 'mysql+pymysql://student:***@columbia-mds.csbmzoea3lu9.us-east-1.rds.amazonaws.com:3306/witchcraft'

3 rows affected.

devilref,devilsystemid,devilid,caseref,devil_type,devil_text,createdby,createdate,lastupdatedby,lastupdatedon
DVA/JO/107,JO,107,C/JO/2846,Male,fearful shape,jhm,2001-07-24T10:25:30,LEM,2002-11-01T14:29:50
DVA/JO/108,JO,108,C/EGD/1817,Male,in green clothing,jhm,2001-07-31T13:33:23,jhm,2002-10-22T12:54:21
DVA/JO/109,JO,109,C/EGD/1862,Male,"large, black man",jhm,2001-07-31T14:43:45,LEM,2002-10-10T11:30:37


In [68]:
# this code = 0 or None suggests providing all rows, the full result is displayed, while 10 rows are for default
%config SqlMagic.displaylimit = None

displaylimit: Value None will be treated as 0 (no limit)

In [72]:
# List unique devil_types
%%sql
SELECT DISTINCT devil_type
FROM devilappearance;

Running query in 'mysql+pymysql://student:***@columbia-mds.csbmzoea3lu9.us-east-1.rds.amazonaws.com:3306/witchcraft'

16 rows affected.

devil_type
Male
Female Fairy
Male Fairy
Animal Devil
Spirit
Ghost
Other Demon
Female
None
Fairy


In [74]:
# b) There is also a little description of the type of the devil sighting in the devil_text column.
# How many of the sightings mention the word "black" in the description?

%%sql
SELECT COUNT(*) AS black_text_count
FROM devilappearance
WHERE devil_text LIKE '%black%';

Running query in 'mysql+pymysql://student:***@columbia-mds.csbmzoea3lu9.us-east-1.rds.amazonaws.com:3306/witchcraft'

1 rows affected.

black_text_count
121


In [80]:
# c) What proportion of the devils (in devil_type) are male?
%%sql
SELECT ROUND((SUM(devil_type = 'Male') / COUNT(*)) *100, 2) AS male_proportion
FROM devilappearance;

Running query in 'mysql+pymysql://student:***@columbia-mds.csbmzoea3lu9.us-east-1.rds.amazonaws.com:3306/witchcraft'

1 rows affected.

male_proportion
63.13


After looking into the sex of all devil types, around 63.13% of them are male.

In [92]:
%config SqlMagic.displaylimit = 10

In [112]:
# d) Using REGEXP, find all devil appearances where the devil_text description contains
# multiple words (hint: look for spaces). How many such descriptions are there?
%%sql
SELECT devil_text AS devil_appearance
FROM devilappearance
WHERE devil_text REGEXP '\s+'
LIMIT 10;

Running query in 'mysql+pymysql://student:***@columbia-mds.csbmzoea3lu9.us-east-1.rds.amazonaws.com:3306/witchcraft'

10 rows affected.

devil_appearance
fearful shape
"little man, in blue, blue hat, black staff"
green clothes/kirtall
black man in black clothes
green clothes/grey hat
in brown clothes
black clothes
in black clothes
in black clothes
a little cuttie fellow with a staff


In [104]:
%%sql
SELECT COUNT(devil_text) AS devil_appearance
FROM devilappearance
WHERE devil_text REGEXP '\s+';

Running query in 'mysql+pymysql://student:***@columbia-mds.csbmzoea3lu9.us-east-1.rds.amazonaws.com:3306/witchcraft'

1 rows affected.

devil_appearance
140


## Q3. The trial
Let's take a look at the information on the trial (table).

In [117]:
%%sql
SELECT *
FROM trial
LIMIT 3;

Running query in 'mysql+pymysql://student:***@columbia-mds.csbmzoea3lu9.us-east-1.rds.amazonaws.com:3306/witchcraft'

3 rows affected.

trialref,trialid,trialsystemid,caseref,trialtype,trial_settlement,trial_parish,trial_presbytery,trial_county,trial_burgh,trial_ngr_letters,trial_ngr_easting,trial_ngr_northing,watching,watchingdate,watchingdate_as_date,arrest,arrestdate,arrestdate_as_date,fled,fleddate,fleddate_as_date,actiondropped,actiondropdate,actiondropdate_as_date,actiondroppedby,actionbody,confrontingsuspects,pretrialnotes,processtype,trialdate,trialdate_as_date,trialplace,female_accusers,male_accusers,high_status,defence,verdict,sentence,cjtorder,cjtdate,cjtdate_as_date,circuit,circuitname,localwithcrep,noreftocentral,trialnotes,execution,executionmethod,executiondate,executiondate_as_date,executionplace,exec_settlement,exec_parish,exec_presbytery,exec_county,exec_burgh,exec_ngr_letters,exec_ngr_easting,exec_ngr_northing,posttrialnotes,createdby,createdate,lastupdatedby,lastupdatedon
T/JO/1,1,JO,C/EGD/2120,2,None,None,Aberdeen,Aberdeen,Aberdeen,None,None,None,0,None,None,0,None,None,0,None,None,0,None,None,None,None,0,None,None,17/2/1597,1997-02-17T00:00:00,Tolbooth,19.0,25.0,0,0,Guilty,Execution,0,None,None,0,None,0,0,None,1,Burn,None,None,None,None,None,None,None,Aberdeen,None,None,None,None,jhm,2001-05-16T15:04:51,jhm,2002-10-24T09:32:53
T/JO/100,100,JO,C/JO/2669,2,None,None,None,None,None,None,None,None,0,None,None,0,None,None,0,None,None,0,None,None,None,None,0,Presbytery notes that none of the accused are subject to the authority of ministers or the presbytery and so cannot attest their depositions. Makghie nominated to witness their confessions before commission can be requested. Note about re-examining the suspects in order to confirm their depositions. Once this had been done commission was to be requested.,None,None,None,None,0.0,0.0,0,0,None,None,0,None,None,0,None,0,0,No trial details.,0,None,None,None,None,None,None,None,None,None,None,None,None,None,jhm,2001-06-14T14:29:49,LEM,2002-11-04T10:07:05
T/JO/1000,1000,JO,C/EGD/1474,2,None,None,None,None,None,None,None,None,0,None,None,0,None,None,0,None,None,0,None,None,None,None,0,Commission issued after accused had confessed to some aspect of witchcraft but PC urged the commissioners not to use torture and to ensure that the accused are mentally and legally competent to stand trial.,None,None,None,None,0.0,0.0,0,0,None,None,0,None,None,0,None,0,0,No details,0,None,None,None,None,None,None,None,None,None,None,None,None,None,jhm,2001-10-11T16:24:03,jhm,2002-10-08T10:56:06


In [123]:
%config SqlMagic.displaylimit = 0

In [124]:
%%sql
DESCRIBE trial;

Running query in 'mysql+pymysql://student:***@columbia-mds.csbmzoea3lu9.us-east-1.rds.amazonaws.com:3306/witchcraft'

65 rows affected.

Field,Type,Null,Key,Default,Extra
trialref,text,YES,,None,
trialid,bigint,YES,,None,
trialsystemid,text,YES,,None,
caseref,text,YES,,None,
trialtype,bigint,YES,,None,
trial_settlement,text,YES,,None,
trial_parish,text,YES,,None,
trial_presbytery,text,YES,,None,
trial_county,text,YES,,None,
trial_burgh,text,YES,,None,


In [125]:
#a) What are the average and maximum numbers of male and female accusers?
%%sql
SELECT
    AVG(male_accusers) AS avg_male_accusers,
    MAX(male_accusers) AS max_male_accusers,
    AVG(female_accusers) AS avg_female_accusers,
    MAX(female_accusers) AS max_female_accusers
FROM trial;

Running query in 'mysql+pymysql://student:***@columbia-mds.csbmzoea3lu9.us-east-1.rds.amazonaws.com:3306/witchcraft'

1 rows affected.

avg_male_accusers,max_male_accusers,avg_female_accusers,max_female_accusers
0.4181988158304768,48.0,0.2559226932668329,27.0


In [134]:
# b) Count the number of sentences by sentence type. List them in a table (in descending order),
# excluding missing values. Rename the column headings to something sensible.
%%sql
SELECT
  sentence AS Sentence_Type,
  COUNT(*) AS N_Sentences
FROM trial
WHERE sentence IS NOT NULL
GROUP BY Sentence_Type
ORDER BY N_Sentences DESC;

Running query in 'mysql+pymysql://student:***@columbia-mds.csbmzoea3lu9.us-east-1.rds.amazonaws.com:3306/witchcraft'

10 rows affected.

Sentence_Type,N_Sentences
Execution,205
Released,52
Banishment,27
Declared Fugitive,11
Excommunicated,6
Put to the horn,2
Hang,1
Branded,1
Prison,1
Public Humiliation,1


In [148]:
# c) Do the number of accusers matter for the verdict? Calculate the total number of accusers (male + female)
# for each trial, then compare the average total number of accusers by the type of verdict.
# Again make sure the table is sorted and the headings make sense.

%%sql
SELECT
  verdict,
  AVG(male_accusers + female_accusers) AS average_total_accusers
FROM trial
GROUP BY verdict
ORDER BY average_total_accusers DESC;

Running query in 'mysql+pymysql://student:***@columbia-mds.csbmzoea3lu9.us-east-1.rds.amazonaws.com:3306/witchcraft'

5 rows affected.

verdict,average_total_accusers
Guilty,5.804255319148936
Not Guilty,4.558139534883721
Not Proven,1.0
Half Guilty,0.2857142857142857
None,0.2013053933356235


The average number of accusers varies significantly depending on the verdict. Trials resulting in a 'Guilty' verdict had the highest average number of total accusers at 5.8.
Trials with a 'Not guilty' verdict showed an average of 4.56 accusers, ranking second.
Trials concluding with a 'None' verdict had the lowest average number of accusers, approximately 0.2, most accusers are being given a clear verdict.

## Q4. Tortured Truth



In [152]:
%%sql
SELECT *
FROM trial
LIMIT 10;

Running query in 'mysql+pymysql://student:***@columbia-mds.csbmzoea3lu9.us-east-1.rds.amazonaws.com:3306/witchcraft'

10 rows affected.

trialref,trialid,trialsystemid,caseref,trialtype,trial_settlement,trial_parish,trial_presbytery,trial_county,trial_burgh,trial_ngr_letters,trial_ngr_easting,trial_ngr_northing,watching,watchingdate,watchingdate_as_date,arrest,arrestdate,arrestdate_as_date,fled,fleddate,fleddate_as_date,actiondropped,actiondropdate,actiondropdate_as_date,actiondroppedby,actionbody,confrontingsuspects,pretrialnotes,processtype,trialdate,trialdate_as_date,trialplace,female_accusers,male_accusers,high_status,defence,verdict,sentence,cjtorder,cjtdate,cjtdate_as_date,circuit,circuitname,localwithcrep,noreftocentral,trialnotes,execution,executionmethod,executiondate,executiondate_as_date,executionplace,exec_settlement,exec_parish,exec_presbytery,exec_county,exec_burgh,exec_ngr_letters,exec_ngr_easting,exec_ngr_northing,posttrialnotes,createdby,createdate,lastupdatedby,lastupdatedon
T/JO/1,1,JO,C/EGD/2120,2,None,None,Aberdeen,Aberdeen,Aberdeen,None,None,None,0,None,None,0,None,None,0,None,None,0,None,None,None,None,0,None,None,17/2/1597,1997-02-17T00:00:00,Tolbooth,19.0,25.0,0,0,Guilty,Execution,0,None,None,0,None,0,0,None,1,Burn,None,None,None,None,None,None,None,Aberdeen,None,None,None,None,jhm,2001-05-16T15:04:51,jhm,2002-10-24T09:32:53
T/JO/100,100,JO,C/JO/2669,2,None,None,None,None,None,None,None,None,0,None,None,0,None,None,0,None,None,0,None,None,None,None,0,Presbytery notes that none of the accused are subject to the authority of ministers or the presbytery and so cannot attest their depositions. Makghie nominated to witness their confessions before commission can be requested. Note about re-examining the suspects in order to confirm their depositions. Once this had been done commission was to be requested.,None,None,None,None,0.0,0.0,0,0,None,None,0,None,None,0,None,0,0,No trial details.,0,None,None,None,None,None,None,None,None,None,None,None,None,None,jhm,2001-06-14T14:29:49,LEM,2002-11-04T10:07:05
T/JO/1000,1000,JO,C/EGD/1474,2,None,None,None,None,None,None,None,None,0,None,None,0,None,None,0,None,None,0,None,None,None,None,0,Commission issued after accused had confessed to some aspect of witchcraft but PC urged the commissioners not to use torture and to ensure that the accused are mentally and legally competent to stand trial.,None,None,None,None,0.0,0.0,0,0,None,None,0,None,None,0,None,0,0,No details,0,None,None,None,None,None,None,None,None,None,None,None,None,None,jhm,2001-10-11T16:24:03,jhm,2002-10-08T10:56:06
T/JO/1001,1001,JO,C/EGD/1558,2,None,None,None,None,None,None,None,None,0,None,None,0,None,None,0,None,None,0,None,None,None,None,0,Commission for trial but the accused was to be legally and mentally competent to stand trial.,None,None,None,None,0.0,0.0,0,0,None,None,0,None,None,0,None,0,0,None,0,None,None,None,None,None,None,None,None,None,None,None,None,None,jhm,2001-10-18T11:54:03,jhm,2001-10-18T11:55:28
T/JO/1002,1002,JO,C/EGD/1681,2,None,None,None,None,None,None,None,None,0,None,None,0,None,None,0,None,None,0,None,None,None,None,0,Commission issued after accused had confessed but PC urged caution about use of torture. The accusedwas to confess voluntarily and be of legal and mental competency to stand trial.,None,None,None,None,0.0,0.0,0,0,None,None,0,None,None,0,None,0,0,No details,0,None,None,None,None,None,None,None,None,None,None,None,None,No details,jhm,2001-10-18T12:23:35,jhm,2002-10-21T16:11:47
T/JO/1003,1003,JO,C/EGD/1680,2,None,None,None,None,None,None,None,None,0,None,None,0,None,None,0,None,None,0,None,None,None,None,0,Commission issued after the accused had confessed.,None,None,None,None,0.0,0.0,0,0,None,None,0,None,None,0,None,0,0,No details,0,None,None,None,None,None,None,None,None,None,None,None,None,No details,jhm,2001-10-18T12:44:59,jhm,2001-10-18T12:46:17
T/JO/1004,1004,JO,C/EGD/1682,2,None,None,None,None,None,None,None,None,0,None,None,0,None,None,0,None,None,0,None,None,None,None,0,Commission issued after the accused had confessed.,None,None,None,None,0.0,0.0,0,0,None,None,0,None,None,0,None,0,0,No detai

In [142]:
%%sql
SELECT *
FROM confession
LIMIT 3;

Running query in 'mysql+pymysql://student:***@columbia-mds.csbmzoea3lu9.us-east-1.rds.amazonaws.com:3306/witchcraft'

3 rows affected.

confessionref,confessionsystemid,confessionid,trialref,centraltrialconfession,confessionrec,confessiondate,confessiondate_as_date,confessionplace,confessionlocation,confessiontext,confessiondetails,confessionret,notes,createdby,createdate,lastupdatedby,lastupdatedon
CF/JO/1,JO,1,T/JO/2,0,1,19/4/1643,2043-04-19T00:00:00,Church,Haddington,0,1,0,None,jhm,2001-05-17T12:11:47,jhm,2001-05-17T12:12:27
CF/JO/100,JO,100,T/JO/180,0,1,19/6/1650,1950-06-19T00:00:00,None,Haddington,0,0,0,None,jhm,2001-06-19T10:11:20,jhm,2001-06-19T10:11:33
CF/JO/101,JO,101,T/JO/181,0,1,19/6/1560,1960-06-19T00:00:00,None,Haddington,0,0,0,None,jhm,2001-06-19T10:13:26,jhm,2001-06-19T10:13:39


In [151]:
# a) Left join the trial and confession tables. For what share of trials does the database record confessions?
# Create a results table with the number of all trials, the number of confessions, and the share of trials with
# confessions recorded.

%%sql
SELECT
  COUNT(DISTINCT T.trialref) AS N_trials,
  COUNT(C.confessionref) AS N_confessions,
  ROUND((COUNT(C.confessionref) / COUNT(DISTINCT T.trialref)) *100, 2) AS Share_of_Trials_with_Confessions_Percent
FROM trial AS T
LEFT JOIN confession AS C
  ON T.trialref = C.trialref;

Running query in 'mysql+pymysql://student:***@columbia-mds.csbmzoea3lu9.us-east-1.rds.amazonaws.com:3306/witchcraft'

1 rows affected.

N_trials,N_confessions,Share_of_Trials_with_Confessions_Percent
3211,941,29.31


This suggests that most trials are not matched with a confession, people who were accused did not confess of doing witchcraft.

In [153]:
# b) Only a small number of trials have records of torture. Is there a higher share of confessions among trials
# with records of torture than trials without such record? Hint: You will need to merge on the confession table
# and use a CASE WHEN statement or WHERE conditions with proper parentheses to categorize trials by whether they
# have torture records.

%%sql
SELECT
  CASE
    WHEN T.pretrialnotes LIKE '%torture%' THEN 'With Torture Record'
    ELSE 'Without Torture Record'
  END AS torture_category,
  COUNT(DISTINCT T.trialref) AS total_trials,
  COUNT(C.confessionref) AS num_confessions,
  ROUND((COUNT(C.confessionref) * 100.0) / COUNT(DISTINCT T.trialref), 2) AS percentage_confessions
FROM trial AS T
LEFT JOIN confession AS C
  ON T.trialref = C.trialref
GROUP BY
  torture_category;


Running query in 'mysql+pymysql://student:***@columbia-mds.csbmzoea3lu9.us-east-1.rds.amazonaws.com:3306/witchcraft'

2 rows affected.

torture_category,total_trials,num_confessions,percentage_confessions
With Torture Record,64,54,84.38
Without Torture Record,3147,887,28.19


Thought of process:
- Start from trial again.

- LEFT JOIN to torture (any row ⇒ “has torture record”).

- LEFT JOIN to confession.

- Use CASE WHEN on the joined torture table to split trials into:

    - “Torture record”

    - “No torture record”

- For each group, compute:

    - total number of trials

    - number with confessions

    - share with confessions

*   The analysis categorized trials into two groups: "With Torture Record" and "Without Torture Record," based on whether the `pretrialnotes` column contained the word 'torture'.
*   For each category, the total number of unique trials, the total number of confessions, and the percentage of confessions relative to trials were calculated.
*   The percentage of confessions was determined by dividing the count of confessions by the count of distinct trials for each category, rounded to two decimal places.

Yes, the confession rate is much higher 84.38% with torture, whereas this rate is much lower when torture was not applied (28.19).